client/txn.go

```diff
@@ -71,5 +174,28 @@ func (t *Txn) Prepare(calls ...Call) {
 func (t *Txn) Flush() error {
 	calls := t.prepared
 	t.prepared = nil
+	if len(calls) == 0 {
+		return nil
+	}
 	return t.kv.Run(calls...)
 }
+
+func (t *Txn) updateReadOnly(calls []Call) {
+	for _, c := range calls {
+		if b, ok := c.Args.(*proto.BatchRequest); ok {
+			for _, br := range b.Requests {
+				t.updateReadOnlyRequest(br.GetValue().(proto.Request))
+			}
+			continue
+		}
+		t.updateReadOnlyRequest(c.Args)
+	}
+}
+
+func (t *Txn) updateReadOnlyRequest(r proto.Request) {
+	if t.readOnly {
+		t.readOnly = !proto.IsTransactional(r)
+	} else if _, ok := r.(*proto.EndTransactionRequest); ok {
+		t.readOnly = true
```

Comment:
How is this clearer? A read/write transaction doesn't become read-only just because it has been explicitly ended; it would be wrong to use this `readOnly` flag for anything but deciding whether to send an EndTransaction.